### DISCUSSION TITLE
How can I properly release microphone and camera resources when a participant hangs up a call in my WebRTC application?

### DISCUSSION THREAD

**Discussion by Mohit-3196** [ant-media/Ant-Media-Server#5251] [FAQs] [answered]
Let's say two participants are on an ongoing call and if any of the parties hangups the call, the client application should properly release the mic/camera. 

1. Client app initializes the `WebRTCAdaptor` instance. As a result, the user is asked to give cam/mic permission.
2. User is on call. If the user uses the Chrome browser, the so-called "recoding" icon is shown on the browser tab.
3. Client app invokes the `leaveFromRoom` method of the WebRTCAdaptor instance.
4. As a result client app receives the `leavedFromRoom` server event in the callback, where removing all remote videos logic is written.
5. After that client app receives `publish_finished`  and `play_finished`  server events in the callback.
6. Client app tries to dispose of the `WebRTCAdaptor`  instance (webRTCAdaptorInstance = null).
Result: camera/mic resources are not released and the "recoding" icon on the Chrome tab remains.

Question: How and when should we release the microphone and camera resources to address this issue?

**Comment by zerostack0** [comment]
From time to time we forget to close all tracks after the call is finished and because of the track is still in use, the camera light won't turn off. You have to stop them like this: 

`localVideo.current.srcObject.getVideoTracks().forEach((track) => track.stop());`

**Comment by mekya** [comment, marked as answer]
Update: 

I think the easiest part to close streams and dispose recording icon, just call 
```
webRTCAdaptor.closeStream()
``` 

If you want to disconnect from websocket and clear all peer connections, just call
```
webRTCAdaptor.closeWebSocket();
```
------------
Another option is to turn off the camera for video stream. It turnoff the camera light. 
```
webRTCAdaptor.turnOffLocalCamera()
```

On the other hand, recording icon is still there. I think it's whether about permission or microphone usage.  I just need to check if it can be disposed completely.

**Reply by mekya** [reply]
I think I'll find a way to release the microphone and update you.

**Reply by Protti19** [reply]
Did you find a way to get rid of the icon ?
